### PR TITLE
feat(@clayui/core): add initial dnd implementation for TreeView

### PR DESCRIPTION
**Pull request by julien**
(W.I.P. Do not merge yet, thanks)

### CODE REVIEW

File: packages/clay-core/src/tree-view/TreeView.tsx [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+import {DndProvider} from 'react-dnd';
+import HTML5Backend from 'react-dnd-html5-backend';
+
+import {ChildrenFunction, Collection, ICollectionProps} from './Collection';
+import {TreeViewGroup} from './TreeViewGroup';
+import {TreeViewItem, TreeViewItemStack} from './TreeViewItem';
+import {Icons, TreeViewContext} from './context';
+import {ITreeProps, useTree} from './useTree';
+
+function generateItemIndices<T>(
```
Comment (julien) [inline packages/clay-core/src/tree-view/TreeView.tsx:17]: @matuzalemsteles I guess that this function could be inside (or part of) a custom hook.

Initially what I had done is something like this (in pseudo code)

```tsx
export function TreeView<T>({
        ...
	items: initialItems,
	...
}: ITreeViewProps<T>) {
   const [items, setItems] = React.useState(generateItemIndices(initialItems));
   ...
}
```

turns out that broke everything (nothing was rendered) and I couldn't figure out why so I went with this "ugly solution" just to have something working. A hook would probably be the good place to also "move" the items which is currently done in `TreeViewItem`. Apart from that, there's not much but I also noticed that when dropping items in some cases there's no update. If you have any ideas let me know.

File: packages/clay-core/src/tree-view/TreeView.tsx [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+	items: Array<T> | Record<string, T> | undefined
+) {
+	if (!items) {
```
Comment (julien) [inline packages/clay-core/src/tree-view/TreeView.tsx:20]: @matuzalemsteles  About this, I know why we defined items as `Array<T> | Record<string, T> | undefined` (we talked about in either #4218 or #4222), but there's got to be a better way to deal with these union types, I need to clean up.

File: packages/clay-core/src/tree-view/TreeView.tsx (hunk 1 above)
Comment (matuzalemsteles) [reply packages/clay-core/src/tree-view/TreeView.tsx:17]: @julien yeah I'm taking a look at it now. Probably as you are manipulating the nested structure, this does not cause a rendering, the ideal would be to create a copy of the structure and then manipulate and save in state, I am thinking of some solutions now.

**Comment by matuzalemsteles** [comment]
Hey @julien I did some exploration today, send a PR #4254 to review and verify some details. See my changes and comments in PR #4254, still very early.

**Comment by julien** [comment]
Closing here since it seems we're going to use #4254

**Pull request closed without merging**